File: cycle_ll.java
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */
public class Solution {
    /*
    Time complexity: O(n)
    Space complexity: O(1)
    */
    
    public ListNode detectCycle(ListNode head) {
        ListNode slow = head;
        ListNode fast = head;
        
        // traverse the list to find cycles
        while(fast != null && fast.next != null){
            slow = slow.next;
            fast = fast.next.next;
            
            // found a cycle
            if(slow == fast){
                ListNode dummy = head;
                // when dummy and slow will be equal then 
                // point where the cycle starts is reached
                while(dummy != slow){
                dummy = dummy.next;
                slow = slow.next;
                }
                // point where the cycle starts
                return slow;
            }
            
        }
        // return null if there is no cycle
        return null;
    }
}

File: cycleLL.cpp
TC:O(N)
SC:O(1)
// Using fast and slow pointer ,move fast at twice the speed of low
// If there is a cycle both meet at some point 

class Solution {
public:
    bool hasCycle(ListNode *head) {
        ListNode* fast=head;
        ListNode* slow=head;

        while(fast && fast->next)
        {
            fast=fast->next->next;
            slow=slow->next;
            if(fast==slow)
            return true;
        }
        return false;
    }
};